src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperExecutionContext.cs

```diff
@@ -171,18 +170,18 @@ public Task ExecuteChildContentAsync()
         /// <returns>A <see cref="Task"/> that on completion returns the rendered child content.</returns>
         /// <remarks>
         /// Child content is only executed once. Successive calls to this method or successive executions of the 
-        /// returned <see cref="Task{string}"/> return a cached result.
+        /// returned <see cref="Task{TagHelperContent}"/> return a cached result.
         /// </remarks>
-        public async Task<string> GetChildContentAsync()
+        public async Task<TagHelperContent> GetChildContentAsync()
         {
             if (_childContent == null)
             {
-                _startWritingScope();
+                _startTagHelperWritingScope();
                 await _executeChildContentAsync();
-                _childContent = _endWritingScope().ToString();
+                _childContent = _endTagHelperWritingScope();
             }
 
-            return _childContent;
+            return new DefaultTagHelperContent().SetContent(_childContent);
```

Comment:
The `ChildContentRetrieved` block calls into the user-backed method `GetChildContentAsync` for two reasons as mentioned in the doc comments: 1. To not re-execute user code and 2. to cache the original result. (Note we should mention that it's now a copied version of the original in the xml doc)

The whole purpose of going from `string` => `TagHelperContent` in this PR is to limit allocations, I feel it's important to do this in one of the core places (if not the primary place) that spawned the purpose for this issue to begin with.